bluehoover/clickhouse.py

```diff
@@ -474,15 +484,70 @@ async def insert_batch(self, batch: list[tuple]) -> None:
             self.client = None  # Force reconnect on next attempt
             raise
 
+class FilesystemCheckpoint:
+    async def get_checkpoint(self) -> int | None:
+        try:
+            with open(CURSOR_FILENAME, "r") as f:
+                return int(f.read())
+        except (FileNotFoundError, ValueError, TypeError, OSError):
+            return None
+
+    async def save_checkpoint(self, cursor: int) -> None:
+        with open(CURSOR_FILENAME, "w") as f:
+            f.write(str(cursor))
+
+class KeepermapDeletingCheckpoint:
+    def __init__(self):
+        self.clickhouse = ClickHouseManager()
+
+    async def get_checkpoint(self) -> int | None:
+        try:
+            client = await self.clickhouse.get_client()
+            result = await client.query("SELECT max(cursor) FROM cursor")
+            return int(result.first_row[0])
+        except (ClickHouseError, ValueError, TypeError) as e:
+            logger.error(f"Error reading checkpoint: {e}", exc_info=true)
+            return None
+
+    async def save_checkpoint(self, cursor: int) -> None:
+        try:
+            client = await self.clickhouse.get_client()
+            insertQuery = f"INSERT INTO TABLE cursor VALUES ({cursor}, {cursor})"
+            deleteQuery = f"DELETE FROM cursor WHERE cursor < {cursor}"
+            await client.command(insertQuery)
+            await client.command(deleteQuery)
+        except Exception as e:
+            logger.error(f"Error saving checkpoint: {e}", exc_info=true)
+
+class KeepermapUpdatingCheckpoint:
+    def __init__(self):
+        self.clickhouse = ClickHouseManager()
+
+    async def get_checkpoint(self) -> int | None:
+        try:
+            client = await self.clickhouse.get_client()
+            result = await client.query("SELECT max(cursor) FROM cursor")
+            return int(result.first_row[0])
+        except (ClickHouseError, ValueError, TypeError) as e:
+            logger.error(f"Error reading checkpoint: {e}", exc_info=true)
+            return None
+
+    async def save_checkpoint(self, cursor: int) -> None:
+        try:
+            client = await self.clickhouse.get_client()
+            insertQuery = f"INSERT INTO TABLE cursor VALUES ('cursor', {cursor})"
+            await client.command(insertQuery)
+        except Exception as e:
+            logger.error(f"Error saving checkpoint: {e}", exc_info=true)
```

Comment:
Now that we're passing `exc_info` suggest:

```suggestion
            logger.error(f"Error saving checkpoint", exc_info=true)
```